firehose/info/endpoint_info.go

```diff
@@ -0,0 +1,200 @@
+package info
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/streamingfast/bstream"
+	"github.com/streamingfast/bstream/hub"
+	pbbstream "github.com/streamingfast/bstream/pb/sf/bstream/v1"
+	"github.com/streamingfast/dstore"
+	pbfirehose "github.com/streamingfast/pbgo/sf/firehose/v2"
+	"go.uber.org/zap"
+)
+
+type InfoServer struct {
+	sync.Mutex
+
+	responseFiller func(block *pbbstream.Block, resp *pbfirehose.InfoResponse) error
+	response       *pbfirehose.InfoResponse
+	ready          chan struct{}
+	initDone       bool
+	initError      error
+	logger         *zap.Logger
+}
+
+func (s *InfoServer) Info(ctx context.Context, request *pbfirehose.InfoRequest) (*pbfirehose.InfoResponse, error) {
+	select {
+	case <-s.ready:
+		return s.response, nil
+	default:
+		return nil, fmt.Errorf("info server not ready")
+	}
+}
+
+func NewInfoServer(
+	chainName string,
+	chainNameAliases []string,
+	blockIDEncoding pbfirehose.InfoResponse_BlockIdEncoding,
+	blockFeatures []string,
+	firstStreamableBlock uint64,
+	responseFiller func(block *pbbstream.Block, resp *pbfirehose.InfoResponse) error,
+	logger *zap.Logger,
+) *InfoServer {
+
+	resp := &pbfirehose.InfoResponse{
+		ChainName:               chainName,
+		ChainNameAliases:        chainNameAliases,
+		BlockIdEncoding:         blockIDEncoding,
+		BlockFeatures:           blockFeatures,
+		FirstStreamableBlockNum: firstStreamableBlock,
+	}
+
+	return &InfoServer{
+		responseFiller: responseFiller,
+		response:       resp,
+		ready:          make(chan struct{}),
+		logger:         logger,
+	}
+}
+
+func validateInfoResponse(resp *pbfirehose.InfoResponse) error {
+	switch {
+	case resp.ChainName == "":
+		return fmt.Errorf("chain name is not set")
+	case resp.BlockIdEncoding == pbfirehose.InfoResponse_BLOCK_ID_ENCODING_UNSET:
+		return fmt.Errorf("block id encoding is not set")
+	case resp.FirstStreamableBlockId == "":
+		return fmt.Errorf("first streamable block id is not set")
+	}
+
+	return nil
+}
+
+// multiple apps (firehose, substreams...) can initialize the same server, we only need one
+func (s *InfoServer) Init(ctx context.Context, fhub *hub.ForkableHub, mergedBlocksStore dstore.Store, oneBlockStore dstore.Store, logger *zap.Logger) error {
+	s.Lock()
+	defer func() {
+		s.initDone = true
+		s.Unlock()
+	}()
+
+	if s.initDone {
+		return s.initError
+	}
+
+	if err := s.init(ctx, fhub, mergedBlocksStore, oneBlockStore, logger); err != nil {
+		s.initError = err
+		return err
+	}
+
+	close(s.ready)
+	return nil
+}
+
+func (s *InfoServer) getBlockFromMergedBlocksStore(ctx context.Context, blockNum uint64, mergedBlocksStore dstore.Store) *pbbstream.Block {
+	for {
+		if ctx.Err() != nil {
+			return nil
+		}
+
+		block, err := bstream.FetchBlockFromMergedBlocksStore(ctx, blockNum, mergedBlocksStore)
+		if err != nil {
+			continue
+		}
+		return block
+	}
```

Comment:
This would require a sleep too, if something is severly broken and `FetchBlockFromMergedBlocksStore` return rapidly an error, this might burn CPUs.